### PR TITLE
Document teamcity pipeline

### LINKED ISSUE #1331: Document TeamCity pipeline
The TeamCity configuration doesn't change too often, so we could dedicate a part of our documentation to describe it.

- Make a mermaid diagram of its pipeline
- Document its function in text form

### PR DESCRIPTION
**Pull request by Jingru923**
Fixes #1331

### CODE REVIEW

File: docs/contribute/ci.qmd [hunk 1]
```diff
@@ -0,0 +1,71 @@
+---
+title: "Continuous integration"
+---
+
+Continuous integration is about commits being merged frequently and automatically. When a new commit is made, a series of tests will be done to make sure that this commit is error-free, appliance format, robust in difference environment and safe to be merged. This process drive each new development through building, testing, quality checking, it is like a pipeline.
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:5]:
```suggestion
Continuous integration is about commits being merged frequently. When a new commit is made, a series of tests will be done to make sure that this commit is error-free and robust in different environments.
```

File: docs/contribute/ci.qmd [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+This page contains an extensive explanation on how the Ribasim pipeline works.
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:7]:
```suggestion
This page contains an extensive explanation on how the Ribasim continuous integration works.
```

File: docs/contribute/ci.qmd [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+# Github actions
+When a new development is to be merged into the main branch, one can either make a pull request or commit to main branch directly. Before a commit is added to main branch, a series of GitHub action will be performed. Github actions includes following checks:
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:10]:
```suggestion
When proposing new changes to the code base a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) is opened.
With [GitHub Actions](https://docs.github.com/en/actions), Github provides their own continuous integration service.
With every commit pushed to the branch, a series of GitHub action jobs will be performed.
They include the following checks:
```

File: docs/contribute/ci.qmd [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+* Julia Run Testmodels: This is to make sure all the test models can be run successfully
+* Julia Tests: Run Julia unit tests on multiple platforms, depends on generated testmodels
+* Docs: Executes our documentation examples
+* Pre-commit: Checks code quality with linters, adjust the format
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:15]:
```suggestion
* Pre-commit: Checks Python code quality with linters and enforces formatting
```

File: docs/contribute/ci.qmd [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+* Python codegen: Check file difference and make sure that code generation wouldn't change any files
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:16]:
```suggestion
* Python codegen: Make sure code generation is till up to date by checking that code generation wouldn't change any files
```

File: docs/contribute/ci.qmd [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+* QGIS Tests: Run QGIS unit tests
+
+```{mermaid}
+graph LR
+    A[New development]-->B[Github actions]
+    B-->C[Merge]
+```
+
+# TeamCity
+Ribasim has another cloud-based CI/CD pipeline that is in the TeamCity. TeamCity closely monitors Github and manage the release process.
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:28]:
```suggestion
Ribasim has another cloud-based CI/CD service based on [TeamCity](https://www.jetbrains.com/teamcity/).
TeamCity interacts with GitHub and manage the release process.
```

File: docs/contribute/ci.qmd [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+```{mermaid}
+graph LR
+    A[TeamCity]-->|Monitoring|B[Github]
+    A-->C(Release)
+```
+
+But TeamCity is not always triggered by every new development. TeamCity is only used or triggered in the following circumustances:
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:36]:
```suggestion
TeamCity is not always triggered by every new development. 
TeamCity is only used or triggered in the following circumustances:
```

File: docs/contribute/ci.qmd [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+* When a action takes too long to run on Github Action
+* When the release depends on the artifects. For example, a new build needs to be publish
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:39]:
```suggestion
* When the release depends on the artifacts.
```

File: docs/contribute/ci.qmd [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    TeamCity constantly monitors and has the information of every commit and when a commit is with a tag starts with `v20`, it trigger the release process.
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:41]:
```suggestion
    TeamCity constantly monitors the GitHub repository. When a tag is added that starts with `v20`, it triggers the release process.
```

File: docs/contribute/ci.qmd [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+* When other TeamCity projects depend on artifacts of Ribasim (e.g. the coupler)
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:43]:
```suggestion
* When other TeamCity projects depend on artifacts of Ribasim (e.g. iMOD Coupler)
```

File: docs/contribute/ci.qmd [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+In the release, we publish the generated testmodels, ribasim cli in Windows and Linux, Ribasim QGIS, and source code
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:45]:
```suggestion
In the release, we include the generated testmodels, Ribasim CLI on Windows and Linux, Ribasim QGIS, and the source code.
```

File: docs/contribute/ci.qmd [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+When TeamCity detects an release tag, the release process will be triggered. TeamCity will start to pack the artifects.
+
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:48]:
```suggestion
```

File: docs/contribute/ci.qmd [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+We have the following pipeline to generate artifects for releasing:
+
+* Generate Testmodels: produces generated_testmodels artifect which is part of release.
+* Make GitHub Release: uses artifacts and it makes release
+* Build libribasim: builds binary of Ribasim in Linux and Windows. The artifact is tested in `Test ribasim_api` and used by iMOD Coupler
+* Build ribasim_cli: builds command line interface in Linux and Windows,its artifact is tested in `Test ribasim_cli` and used by release
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:54]:
```suggestion
* Build ribasim_cli: builds command line interface in Linux and Windows, its artifact is tested in `Test ribasim_cli` and used by release
```

File: docs/contribute/ci.qmd [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+* Test ribasim_api: tests libribasim artifact in Linux and Windows
+* Test ribasim_cli: tests ribasim_cli artifact in Linux and Windows
+
+::: {.callout-note}
+Make Github Release does not publish Test libribasim and Test ribasim_cli. It only publish artifect of Build libribasim and Build ribasim_cli if they past their tests.
```
Comment (Hofer-Julian) [inline docs/contribute/ci.qmd:59]:
```suggestion
Make Github Release does not publish artifacts of "Test libribasim" and "Test ribasim_cli".
It only publishes artifacts of "Build libribasim" and "Build ribasim_cli" if the beforementioned tests pass.
```

**Review by Hofer-Julian — CHANGES_REQUESTED** [review]
I took a quick look and suggested a few things.

One thing that would make reviewing easier is if you would make a newline after every sentence. This doesn't change the produced output, but makes it easier to comment.

**Review by Hofer-Julian — APPROVED** [review]

**Comment by evetion** [comment]
I've got a few more changes, just hold off merging please :)

**Comment by evetion** [comment]
Nvm 😓

**Comment by Jingru923** [comment]
@evetion I clicked merge like 5 seconds before seeing your comments. Can you put your comments here and I will do a follow-up PR?

**Pull request merged**